### PR TITLE
Refactor Lexer

### PR DESCRIPTION
**Pull request by chenyan-dfinity**
* Use `logos` lexer to replace the hard-coded state machine.
* Support float e notation
* Support nested comments
* Experiment with `codespan-reporting` in CLI

### CODE REVIEW

File: tools/didc/src/main.rs [hunk 1]
```diff
@@ -115,11 +117,34 @@ fn check_file(env: &mut TypeEnv, file: &Path) -> candid::Result<Option<Type>> {
...
+    }
+}
+
+fn main() -> Result<()> {
+    let writer = StandardStream::stderr(term::termcolor::ColorChoice::Auto);
+    let config = term::Config::default();
+
     match Command::from_args() {
         Command::Check { input } => {
+            let file = SimpleFile::new(input.to_str().unwrap(), std::fs::read_to_string(&input)?);
```
Comment (kritzcreek) [inline tools/didc/src/main.rs:139]: Without an actual filemap this doesn't do much. Did you want to merge this experiment, or should this be its own PR?

File: rust/candid/src/parser/token.rs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    StartString,
+    // This token is not derived. Stores the unescaped string
+    Text(String),
+    #[regex("[+-]", |lex| lex.slice().chars().next())]
+    Sign(char),
+    #[regex("[0-9][_0-9]*", parse_number)]
+    Decimal(String),
+    #[regex("0[xX][0-9a-fA-F][_0-9a-fA-F]*", parse_number)]
+    Hex(String),
+    #[regex("[0-9]*\\.[0-9]*", parse_number)]
```
Comment (kritzcreek) [inline rust/candid/src/parser/token.rs:75]: This'll accept leading zeroes, do we want that?

**Review by kritzcreek — APPROVED** [review]
Nice! Any idea how the compile times look?

File: rust/candid/src/parser/token.rs (hunk 2 above)
Comment (chenyan-dfinity) [reply rust/candid/src/parser/token.rs:75]: This is the same as the Motoko lexer, and the spec says it's should be the same as Motoko :)

File: tools/didc/src/main.rs (hunk 1 above)
Comment (chenyan-dfinity) [reply tools/didc/src/main.rs:139]: I think it's fine to merge it, as it doesn't touch the main candid crate. 

I actually don't quite like the filemap idea, it seems an overkill to store all files we ever touched in memory, just for reporting a nice error message. But I don't have better ideas neither.

**Comment by chenyan-dfinity** [comment]
> Any idea how the compile times look?

I didn't notice any compile time increase. Admittedly, `lalrpop` takes a lot of time to compile, so other increases all seem negligible :)
